File: src/main/java/no/fintlabs/validation/ValidationErrorsFormattingService.java
package no.fintlabs.validation;

import org.springframework.stereotype.Service;
import org.springframework.web.bind.support.WebExchangeBindException;

import jakarta.validation.ConstraintViolation;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class ValidationErrorsFormattingService {

    public <T> String format(Set<ConstraintViolation<T>> errors) {
        return "Validation error" + (errors.size() > 1 ? "s:" : ":") + " " +
               errors
                       .stream()
                       .map(constraintViolation -> "'" + (
                                       constraintViolation.getPropertyPath().toString().isBlank()
                                               ? ""
                                               : constraintViolation.getPropertyPath().toString() + " "
                               ) + constraintViolation.getMessage() + "'"
                       )
                       .sorted(String::compareTo)
                       .collect(Collectors.joining(", ", "[", "]"));
    }

    public <T> String format(WebExchangeBindException e) {
        return e.getReason() + (e.getFieldErrorCount() > 1 ? "s:" : ":") + " " +
               e.getFieldErrors()
                       .stream()
                       .map(error -> "value '" + error.getRejectedValue() +
                                     "' for field '" + error.getField() +
                                     "' contains invalid invalid type or format")
                       .sorted(String::compareTo)
                       .collect(Collectors.joining(", ", "[", "]"));
    }

}

File: src/main/kotlin/no/novari/flyt/history/validation/ValidationErrorsFormattingService.kt
package no.novari.flyt.history.validation

import jakarta.validation.ConstraintViolation
import org.springframework.stereotype.Service
import org.springframework.validation.BindException

@Service
class ValidationErrorsFormattingService {
    fun <T> format(errors: Set<ConstraintViolation<T>>): String {
        val prefix = if (errors.size > 1) "Validation errors:" else "Validation error:"
        val formattedErrors =
            errors
                .asSequence()
                .map { violation ->
                    val propertyPath = violation.propertyPath.toString()
                    val propertyPrefix = if (propertyPath.isBlank()) "" else "$propertyPath "
                    "'$propertyPrefix${violation.message}'"
                }.sorted()
                .joinToString(prefix = "[", postfix = "]")

        return "$prefix $formattedErrors"
    }

    fun format(exception: BindException): String {
        val prefix = if (exception.fieldErrorCount > 1) "Validation errors:" else "Validation error:"
        val formattedErrors =
            exception.fieldErrors
                .asSequence()
                .map { error ->
                    "value '${error.rejectedValue}' for field '${error.field}' contains invalid invalid type or format"
                }.sorted()
                .joinToString(prefix = "[", postfix = "]")

        return "$prefix $formattedErrors"
    }
}
